### PR TITLE
planner: distinguish FullScan and the RangeScan in explain/desc info

### LINKED ISSUE #14601: Distinguish FullScan and the RangeScan more clearly.
This is one advice from DBAs.

The following one is the current explain result.
```
mysql> explain select * from t;
+-------------------+----------+-----------+------------------------------------------------------------+
| id                | count    | task      | operator info                                              |
+-------------------+----------+-----------+------------------------------------------------------------+
| TableReader_5     | 10000.00 | root      | data:TableScan_4                                           |
| └─TableScan_4     | 10000.00 | cop[tikv] | table:t, range:[-inf,+inf], keep order:false, stats:pseudo |
+-------------------+----------+-----------+------------------------------------------------------------+
2 rows in set (0.00 sec)
mysql> explain select * from t where a > 1;
+-------------------+---------+-----------+---------------------------------------------------------+
| id                | count   | task      | operator info                                           |
+-------------------+---------+-----------+---------------------------------------------------------+
| TableReader_6     | 3333.33 | root      | data:TableScan_5                                        |
| └─TableScan_5     | 3333.33 | cop[tikv] | table:t, range:(1,+inf], keep order:false, stats:pseudo |
+-------------------+---------+-----------+---------------------------------------------------------+
2 rows in set (0.00 sec)

```
The advice is that 
- changing the `TableScan` to `TableFullScan` and don't output the range info in `operator info` if it's a full table scan
- changing the `TableScan` to `TableRangeScan` if it's not a full scan.
- For `IndexReader`, it's the same.
- For `IndexLookUpReader`, the `IndexScan` is the same with others, and we just change the `TableScan` to `TableRowIDScan` to make its meaning more clearly.

### PR DESCRIPTION
**Pull request by hsqlu**
### What problem does this PR solve? 

fix #14601 

### What is changed and how it works?

Implement ExplainID() for PhysicalIndexScan and PhysicalTableScan.

### Check List 
Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change 

```
func (ran *Range) IsFullRange() bool 
```

### CODE REVIEW

File: util/ranger/types.go [hunk 1]
```diff
@@ -77,6 +77,22 @@ func (ran *Range) IsPoint(sc *stmtctx.StatementContext) bool {
...
 
+//IsFullRange check if the range is full scan range
+func (ran *Range) IsFullRange() bool {
+	if len(ran.LowVal) != len(ran.HighVal) {
+		return false
+	}
+	for i := range ran.LowVal {
+		a := ran.LowVal[i]
+		b := ran.HighVal[i]
+		if ("-inf" != formatDatum(a, true) && "NULL" != formatDatum(a, true)) ||
```
Comment (lamxTyler) [inline util/ranger/types.go]: Use a variable for the result?

File: util/ranger/types.go [hunk 2]
```diff
@@ -77,6 +77,22 @@ func (ran *Range) IsPoint(sc *stmtctx.StatementContext) bool {
...
+			("+inf" != formatDatum(b, false) && "NULL" != formatDatum(a, true)) {
```
Comment (lamxTyler) [inline util/ranger/types.go]: `formatDatum(b, false)`

File: planner/core/task.go [hunk 3]
```diff
@@ -649,6 +649,9 @@ func finishCopTask(ctx sessionctx.Context, task task) task {
 			indexPlan:      t.indexPlan,
 			ExtraHandleCol: t.extraHandleCol,
 		}.Init(ctx, t.tablePlan.SelectBlockOffset())
+		if tp, ok := p.tablePlan.(*PhysicalTableScan); ok {
```
Comment (francis0407) [inline planner/core/task.go]: What if the `tablePlan` is `Selection->TableScan`?

Comment (hsqlu) [reply planner/core/task.go]: I didn't take this scenario into account, will fix it.

File: planner/core/task.go [hunk 4]
```diff
@@ -716,6 +730,15 @@ func finishCopTask(ctx sessionctx.Context, task task) task {
...
 }
 
+// updateTableScanChild is to update the isChildOfIndexLookUp attribute of PhysicalTableScan child
+func (sel *PhysicalSelection) updateTableScanChild() {
+	for _, child := range sel.children {
+		if ts, ok := child.(*PhysicalTableScan); ok {
+			ts.SetIsChildOfIndexLookUp(true)
+		}
+	}
+}
```
Comment (francis0407) [inline planner/core/task.go]: Could we use just one function to set it?
```golang
func SetTableScanToTableRowIDScan(p PhysicalPlan) {
    if ts, ok := p.(*PhysicalTableScan); ok {
       ts.SetIsChildOfIndexLoopUp(true)
    } else {
        for _, child := range p.Children() {
            SetTableScanToTableRowIDScan(child)
        }
    }
}
```

File: util/ranger/types_test.go [hunk 5]
```diff
@@ -18,6 +18,7 @@ import (
 	"github.com/pingcap/tidb/sessionctx/stmtctx"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/util/ranger"
+	"math"
```
Comment (francis0407) [inline util/ranger/types_test.go]: move this to the top of the import list.

**Comment by hsqlu** [comment]
> LGTM
@lzmhhh123 Updated task.go. PTAL. Thanks

File: util/ranger/types_test.go (hunk 5 above)
Comment (francis0407) [reply util/ranger/types_test.go]: Please move this to the top of the import list, like:
```golang
import (
    "math"

    "github.com/pingcap/..."
)
```

File: planner/core/explain.go [hunk 6]
```diff
@@ -27,13 +28,24 @@ import (
...
 	"github.com/pingcap/tidb/util/set"
+	"github.com/pingcap/tidb/util/stringutil"
 )
 
 // ExplainInfo implements Plan interface.
 func (p *PhysicalLock) ExplainInfo() string {
 	return p.Lock.String()
 }
 
+//ExplainID override the ExplainID in order to match different range
```
Comment (francis0407) [inline planner/core/explain.go]:
```suggestion
// ExplainID overrides the ExplainID in order to match different range.
```

File: planner/core/explain.go [hunk 7]
```diff
@@ -99,11 +105,44 @@ func (p *PhysicalIndexScan) explainInfo(normalized bool) string {
...
+	}
+	return true
+}
+
 // ExplainNormalizedInfo implements Plan interface.
 func (p *PhysicalIndexScan) ExplainNormalizedInfo() string {
 	return p.explainInfo(true)
 }
 
+//ExplainID override the ExplainID in order to match different range
```
Comment (francis0407) [inline planner/core/explain.go]: ditto

**Comment by hsqlu** [comment]
> The rest LGTM.

@francis0407 @lzmhhh123   PTAL.

File: planner/core/explain.go [hunk 8]
```diff
@@ -99,11 +105,44 @@ func (p *PhysicalIndexScan) explainInfo(normalized bool) string {
...
 func (p *PhysicalIndexScan) ExplainNormalizedInfo() string {
 	return p.explainInfo(true)
 }
 
+// ExplainID overrides the ExplainID in order to match different range.
+func (p *PhysicalTableScan) ExplainID() fmt.Stringer {
+	return stringutil.MemoizeStr(func() string {
+		if p.isChildOfIndexLookUp {
+			return "TableRowIDScan_" + strconv.Itoa(p.id)
+		} else if len(p.Ranges) > 0 && p.isFullScan() {
```
Comment (zz-jason) [inline planner/core/explain.go]: How about 
```suggestion
		} else if p.isFullScan() {
```

File: cmd/explaintest/r/explain_indexmerge.result [hunk 9]
```diff
@@ -8,61 +8,61 @@ explain select * from t where a < 50 or b < 50;
...
 id	count	task	operator info
 IndexMerge_12	147.00	root	
-├─TableScan_8	49.00	cop[tikv]	table:t, range:[-inf,50), keep order:false
-├─IndexScan_9	49.00	cop[tikv]	table:t, index:b, range:[-inf,50), keep order:false
-├─IndexScan_10	49.00	cop[tikv]	table:t, index:c, range:[-inf,50), keep order:false
-└─TableScan_11	147.00	cop[tikv]	table:t, keep order:false
+├─TableRangeScan_8	49.00	cop[tikv]	table:t, range:[-inf,50), keep order:false
+├─IndexRangeScan_9	49.00	cop[tikv]	table:t, index:b, range:[-inf,50), keep order:false
+├─IndexRangeScan_10	49.00	cop[tikv]	table:t, index:c, range:[-inf,50), keep order:false
+└─TableFullScan_11	147.00	cop[tikv]	table:t, keep order:false
```
Comment (zz-jason) [inline cmd/explaintest/r/explain_indexmerge.result]: It should be `TableRowIDScan`. Besides `IndexLookup`, we also need to consider `IndexMerge`

**Pull request merged**